File: Paintroid/src/org/catrobat/paintroid/tools/implementation/BaseToolWithShape.java
/**
 *  Catroid: An on-device visual programming system for Android devices
 *  Copyright (C) 2010-2012 The Catrobat Team
 *  (<http://developer.catrobat.org/credits>)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  http://www.catroid.org/catroid/licenseadditionalterm
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.paintroid.tools.implementation;

import org.catrobat.paintroid.PaintroidApplication;
import org.catrobat.paintroid.R;
import org.catrobat.paintroid.tools.ToolType;
import org.catrobat.paintroid.tools.ToolWithShape;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PointF;
import android.view.Display;
import android.view.WindowManager;

public abstract class BaseToolWithShape extends BaseTool implements
		ToolWithShape {

	protected int mPrimaryShapeColor = PaintroidApplication.applicationContext
			.getResources().getColor(R.color.rectangle_primary_color);
	protected int mSecondaryShapeColor = PaintroidApplication.applicationContext
			.getResources().getColor(R.color.rectangle_secondary_color);
	protected PointF mToolPosition;
	protected Paint mLinePaint;

	public BaseToolWithShape(Context context, ToolType toolType) {
		super(context, toolType);
		Display display = ((WindowManager) context
				.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay();
		mToolPosition = new PointF(display.getWidth() / 2f,
				display.getHeight() / 2f);
		PaintroidApplication.perspective
				.convertFromScreenToCanvas(mToolPosition);
		mLinePaint = new Paint();
		mLinePaint.setColor(mPrimaryShapeColor);
	}

	@Override
	public abstract void drawShape(Canvas canvas);

	protected float getStrokeWidthForZoom(float defaultStrokeWidth,
			float minStrokeWidth, float maxStrokeWidth) {
		float displayScale = mContext.getResources().getDisplayMetrics().density;
		float strokeWidth = (defaultStrokeWidth * displayScale)
				/ PaintroidApplication.perspective.getScale();
		if (strokeWidth < minStrokeWidth) {
			strokeWidth = minStrokeWidth;
		} else if (strokeWidth > maxStrokeWidth) {
			strokeWidth = maxStrokeWidth;
		}
		return strokeWidth;
	}

	protected float getInverselyProportionalSizeForZoom(float defaultSize) {
		float displayScale = mContext.getResources().getDisplayMetrics().density;
		float applicationScale = PaintroidApplication.perspective
				.getScale();
		return (defaultSize * displayScale) / applicationScale;
	}

}

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/BaseToolWithShape.kt
/*
 * Paintroid: An image manipulation application for Android.
 * Copyright (C) 2010-2021 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.catrobat.paintroid.tools.implementation

import android.annotation.SuppressLint
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Point
import android.graphics.PointF
import android.os.Bundle
import android.util.DisplayMetrics
import androidx.annotation.VisibleForTesting
import org.catrobat.paintroid.R
import org.catrobat.paintroid.command.CommandManager
import org.catrobat.paintroid.tools.ContextCallback
import org.catrobat.paintroid.tools.ToolPaint
import org.catrobat.paintroid.tools.Workspace
import org.catrobat.paintroid.tools.options.ToolOptionsViewController
import kotlin.math.max
import kotlin.math.min

private const val BUNDLE_TOOL_POSITION_Y = "TOOL_POSITION_Y"
private const val BUNDLE_TOOL_POSITION_X = "TOOL_POSITION_X"

abstract class BaseToolWithShape @SuppressLint("VisibleForTests") constructor(
    contextCallback: ContextCallback,
    toolOptionsViewController: ToolOptionsViewController,
    toolPaint: ToolPaint,
    workspace: Workspace,
    commandManager: CommandManager
) : BaseTool(
    contextCallback,
    toolOptionsViewController,
    toolPaint,
    workspace,
    commandManager
) {

    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
    @JvmField
    val toolPosition: PointF

    @JvmField
    var primaryShapeColor: Int =
        contextCallback.getColor(R.color.pocketpaint_main_rectangle_tool_primary_color)

    @JvmField
    var secondaryShapeColor: Int = contextCallback.getColor(R.color.pocketpaint_colorAccent)

    @JvmField
    val linePaint: Paint

    @JvmField
    val metrics: DisplayMetrics = contextCallback.displayMetrics

    init {
        val perspective = workspace.perspective
        toolPosition = if (perspective.scale > 1) {
            PointF(
                perspective.surfaceCenterX - perspective.surfaceTranslationX,
                perspective.surfaceCenterY - perspective.surfaceTranslationY
            )
        } else {
            PointF(workspace.width / 2f, workspace.height / 2f)
        }
        linePaint = Paint()
        linePaint.color = primaryShapeColor
    }

    abstract fun drawShape(canvas: Canvas)

    fun getStrokeWidthForZoom(
        defaultStrokeWidth: Float,
        minStrokeWidth: Float,
        maxStrokeWidth: Float
    ): Float {
        val strokeWidth = defaultStrokeWidth * metrics.density / workspace.scale
        return min(maxStrokeWidth, max(minStrokeWidth, strokeWidth))
    }

    fun getInverselyProportionalSizeForZoom(defaultSize: Float): Float {
        val applicationScale = workspace.scale
        return defaultSize * metrics.density / applicationScale
    }

    override fun onSaveInstanceState(bundle: Bundle?) {
        super.onSaveInstanceState(bundle)
        bundle?.apply {
            putFloat(BUNDLE_TOOL_POSITION_X, toolPosition.x)
            putFloat(BUNDLE_TOOL_POSITION_Y, toolPosition.y)
        }
    }

    override fun onRestoreInstanceState(bundle: Bundle?) {
        super.onRestoreInstanceState(bundle)
        bundle?.apply {
            toolPosition.x = getFloat(BUNDLE_TOOL_POSITION_X, toolPosition.x)
            toolPosition.y = getFloat(BUNDLE_TOOL_POSITION_Y, toolPosition.y)
        }
    }

    override fun getAutoScrollDirection(
        pointX: Float,
        pointY: Float,
        screenWidth: Int,
        screenHeight: Int
    ): Point {
        val surfaceToolPosition = workspace.getSurfacePointFromCanvasPoint(toolPosition)
        return scrollBehavior.getScrollDirection(
            surfaceToolPosition.x,
            surfaceToolPosition.y,
            screenWidth,
            screenHeight
        )
    }

    abstract fun onClickOnButton()

    protected open fun drawToolSpecifics(canvas: Canvas, boxWidth: Float, boxHeight: Float) {}
}
